### PR TITLE
feat: bCake V3

### PR DESCRIPTION
**Pull request by Chef-Yogi**
### Summary
🚀🍰🛠️

This pull request adds support for the bCakeFarmBoosterV3 contract, which allows users to boost their yield farming rewards by locking CAKE tokens. It updates the useContract hook and the contract helpers file to import and use the contract address, ABI, and type. It also adds a new file with custom hooks for fetching the whitelist status and activating or deactivating the boost for each farm.

> _Boost your farm rewards_
> _With `bCakeFarmBoosterV3`_
> _Contract hooks added_

### Walkthrough
*  Add custom hooks for interacting with the bCakeFarmBoosterV3 contract ([link](https://github.com/pancakeswap/pancake-frontend/pull/6873/files?diff=unified&w=0#diff-9d9892f9e39d2bf4ca3adf284a224f97f2c289bb33c1d1419c2a057ae453c138R1-R53))
* Add the ABI and address of the bCakeFarmBoosterV3 contract to the config file ([link](https://github.com/pancakeswap/pancake-frontend/pull/6873/files?diff=unified&w=0#diff-448aac5e17df0c0d8f364c1842c542ec2c35969ffefdbf8fe6f05b1404b1bfd8R1-R384), [link](https://github.com/pancakeswap/pancake-frontend/pull/6873/files?diff=unified&w=0#diff-6ccab19ffbec70686dc2221f7b79432ece4302c36067c91063d258209c09ec07R157-R160))
* Add helper functions to get the bCakeFarmBoosterV3 contract instance and address ([link](https://github.com/pancakeswap/pancake-frontend/pull/6873/files?diff=unified&w=0#diff-05729caa7bf80ee8f359513f222175a79f3cd3e8aa704d68f0d5c67f8b010a23R101-R104), [link](https://github.com/pancakeswap/pancake-frontend/pull/6873/files?diff=unified&w=0#diff-dbcbcd450dce97f74ff9b8fff28d36fc457775363d349db2a45e504060c0754aR297-R304))
   * use the getAddress function to validate and checksum the address ([link](https://github.com/pancakeswap/pancake-frontend/pull/6873/files?diff=unified&w=0#diff-05729caa7bf80ee8f359513f222175a79f3cd3e8aa704d68f0d5c67f8b010a23R101-R104))
   * use the getContract function with the bCakeFarmBoosterV3 ABI and address ([link](https://github.com/pancakeswap/pancake-frontend/pull/6873/files?diff=unified&w=0#diff-dbcbcd450dce97f74ff9b8fff28d36fc457775363d349db2a45e504060c0754aR297-R304))
   * cast the result as the BCakeFarmBoosterV3 type ([link](https://github.com/pancakeswap/pancake-frontend/pull/6873/files?diff=unified&w=0#diff-dbcbcd450dce97f74ff9b8fff28d36fc457775363d349db2a45e504060c0754aR297-R304))
* Import the helper functions and the BCakeFarmBoosterV3 type to the useContract hook file ([link](https://github.com/pancakeswap/pancake-frontend/pull/6873/files?diff=unified&w=0#diff-0189a20b061b4c4cd5fd74abb3ff0a22a05b9f5adde9583eb11606cb589a5af4R24), [link](https://github.com/pancakeswap/pancake-frontend/pull/6873/files?diff=unified&w=0#diff-dbcbcd450dce97f74ff9b8fff28d36fc457775363d349db2a45e504060c0754aR30), [link](https://github.com/pancakeswap/pancake-frontend/pull/6873/files?diff=unified&w=0#diff-dbcbcd450dce97f74ff9b8fff28d36fc457775363d349db2a45e504060c0754aR75), [link](https://github.com/pancakeswap/pancake-frontend/pull/6873/files?diff=unified&w=0#diff-dbcbcd450dce97f74ff9b8fff28d36fc457775363d349db2a45e504060c0754aR123))
* Add a custom hook to get the bCakeFarmBoosterV3 contract instance with a signer if possible ([link](https://github.com/pancakeswap/pancake-frontend/pull/6873/files?diff=unified&w=0#diff-0189a20b061b4c4cd5fd74abb3ff0a22a05b9f5adde9583eb11606cb589a5af4R336-R340))
   * use the useMemo hook to memoize the contract instance ([link](https://github.com/pancakeswap/pancake-frontend/pull/6873/files?diff=unified&w=0#diff-0189a20b061b4c4cd5fd74abb3ff0a22a05b9f5adde9583eb11606cb589a5af4R336-R340))
* Fix minor bugs in the useTokenContract, useBytes32TokenContract, usePairContract, and useV3NFTPositionManagerContract hooks ([link](https://github.com/pancakeswap/pancake-frontend/pull/6873/files?diff=unified&w=0#diff-0189a20b061b4c4cd5fd74abb3ff0a22a05b9f5adde9583eb11606cb589a5af4L282-R296), [link](https://github.com/pancakeswap/pancake-frontend/pull/6873/files?diff=unified&w=0#diff-0189a20b061b4c4cd5fd74abb3ff0a22a05b9f5adde9583eb11606cb589a5af4L302-R308), [link](https://github.com/pancakeswap/pancake-frontend/pull/6873/files?diff=unified&w=0#diff-0189a20b061b4c4cd5fd74abb3ff0a22a05b9f5adde9583eb11606cb589a5af4L383-R389))
   * use the correct syntax for optional and typed parameters ([link](https://github.com/pancakeswap/pancake-frontend/pull/6873/files?diff=unified&w=0#diff-0189a20b061b4c4cd5fd74abb3ff0a22a05b9f5adde9583eb11606cb589a5af4L282-R296), [link](https://github.com/pancakeswap/pancake-frontend/pull/6873/files?diff=unified&w=0#diff-0189a20b061b4c4cd5fd74abb3ff0a22a05b9f5adde9583eb11606cb589a5af4L302-R308), [link](https://github.com/pancakeswap/pancake-frontend/pull/6873/files?diff=unified&w=0#diff-0189a20b061b4c4cd5fd74abb3ff0a22a05b9f5adde9583eb11606cb589a5af4L383-R389))
   * use a Boolean wrapper for the withSignerIfPossible parameter ([link](https://github.com/pancakeswap/pancake-frontend/pull/6873/files?diff=unified&w=0#diff-0189a20b061b4c4cd5fd74abb3ff0a22a05b9f5adde9583eb11606cb589a5af4L282-R296), [link](https://github.com/pancakeswap/pancake-frontend/pull/6873/files?diff=unified&w=0#diff-0189a20b061b4c4cd5fd74abb3ff0a22a05b9f5adde9583eb11606cb589a5af4L302-R308), [link](https://github.com/pancakeswap/pancake-frontend/pull/6873/files?diff=unified&w=0#diff-0189a20b061b4c4cd5fd74abb3ff0a22a05b9f5adde9583eb11606cb589a5af4L383-R389))
   * use a default value of ChainId.BSC for the chainId variable ([link](https://github.com/pancakeswap/pancake-frontend/pull/6873/files?diff=unified&w=0#diff-0189a20b061b4c4cd5fd74abb3ff0a22a05b9f5adde9583eb11606cb589a5af4L282-R296))

### CODE REVIEW

File: packages/pools/src/constants/pools/97.ts [hunk 1]
```diff
@@ -54,7 +63,7 @@ export const livePools: SerializedPool[] = [
   },
 ].map((p) => ({
   ...p,
-  contractAddress: getAddress(p.contractAddress),
+  contractAddress: getAddress(p.contractAddress, 97),
```
Comment (ChefMomota) [inline packages/pools/src/constants/pools/97.ts:66]: `97` change it to `ChainId.BSC_TESTNET`

File: packages/utils/isPositionOutOfRange.ts [hunk 2]
```diff
@@ -2,7 +2,10 @@ export function isPositionOutOfRange(
   tickCurrent?: number,
   position?: { tickLower?: number; tickUpper?: number },
 ): boolean {
-  return tickCurrent && position && position.tickLower && position.tickUpper
+  return (tickCurrent || tickCurrent === 0) &&
```
Comment (chefjackson) [inline packages/utils/isPositionOutOfRange.ts:5]: let's use `typeof` for type guard

File: packages/uikit/src/widgets/RoiCalculator/RoiCalculator.tsx [hunk 3]
```diff
@@ -110,6 +112,7 @@ export function RoiCalculator({
   onPriceSpanChange,
   allowApply = false,
   onApply,
+  bCakeBoostedMultiplier = 1,
```
Comment (chefjackson) [inline packages/uikit/src/widgets/RoiCalculator/RoiCalculator.tsx:115]: Can we just update the `cakeAprFactor` so we don't need to add another prop?

**Comment by Chef-Yogi** [comment]
@chefjackson  Done

File: packages/uikit/src/widgets/RoiCalculator/RoiCalculator.tsx [hunk 4]
```diff
@@ -3,6 +3,7 @@ import { FeeAmount, FeeCalculator, Tick, TickMath, sqrtRatioX96ToPrice } from "@
 import { useTranslation } from "@pancakeswap/localization";
 import { useCallback, useMemo, useState } from "react";
 import BigNumber from "bignumber.js";
+import _toNumber from "lodash/toNumber";
```
Comment (chefjackson) [inline packages/uikit/src/widgets/RoiCalculator/RoiCalculator.tsx:6]: Shall we remove this import?

Comment (chefjackson) [inline packages/uikit/src/widgets/RoiCalculator/RoiCalculator.tsx:6]: Shall we remove this import?

**Comment by Chef-Yogi** [comment]
@chefjackson Done

**Review by chefjackson — APPROVED** [review]

**Pull request merged**